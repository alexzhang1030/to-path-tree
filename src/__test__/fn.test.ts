import { pathToTree, walkPathTree } from '..'
import input from './fixtures/input.json'

describe('test', () => {
  it('paths to tree', () => {
    expect(pathToTree(input)).toMatchSnapshot()
  })
  it('walker', () => {
    const tree = pathToTree(input)
    walkPathTree(tree, (node) => {
      for (const item of node.items) {
        if (item.path === 'src/a/b/d/index.ts')
          expect(node.parent?.items).toMatchSnapshot()
      }
    })
  })
})

it('should add directories', () => {
  const tree = pathToTree([
    '/a/',
    '/a/b/',
  ])
  expect(tree).toMatchSnapshot()
})

it('should handle files without extensions like .gitignore', () => {
  const tree = pathToTree([
    '.gitignore',
    '.env',
    'README.md',
    'package.json',
  ])

  // Check that .gitignore is properly parsed
  const gitignoreItem = tree.items.find(item => item.path === '.gitignore')
  expect(gitignoreItem).toBeDefined()
  expect(gitignoreItem?.filename).toBe('.gitignore')
  expect(gitignoreItem?.ext).toBe('')
  expect(gitignoreItem?.isEntry).toBe(false)

  // Check that .env is properly parsed
  const envItem = tree.items.find(item => item.path === '.env')
  expect(envItem).toBeDefined()
  expect(envItem?.filename).toBe('.env')
  expect(envItem?.ext).toBe('')
  expect(envItem?.isEntry).toBe(false)

  // Check that README.md is properly parsed
  const readmeItem = tree.items.find(item => item.path === 'README.md')
  expect(readmeItem).toBeDefined()
  expect(readmeItem?.filename).toBe('README')
  expect(readmeItem?.ext).toBe('md')
  expect(readmeItem?.isEntry).toBe(false)

  // Check that package.json is properly parsed
  const packageItem = tree.items.find(item => item.path === 'package.json')
  expect(packageItem).toBeDefined()
  expect(packageItem?.filename).toBe('package')
  expect(packageItem?.ext).toBe('json')
  expect(packageItem?.isEntry).toBe(false)
})
