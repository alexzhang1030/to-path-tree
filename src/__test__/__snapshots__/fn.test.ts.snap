// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should add directories 1`] = `
{
  "items": [],
  "name": "root",
  "path": "/",
  "relativePath": "/",
  "relativePathName": "",
  "subDirectory": {
    "a": {
      "items": [],
      "name": "a",
      "parent": [Circular],
      "path": "/a/",
      "relativePath": "a/",
      "relativePathName": "a",
      "subDirectory": {
        "b": {
          "items": [],
          "name": "b",
          "parent": [Circular],
          "path": "/a/b/",
          "relativePath": "b/",
          "relativePathName": "b",
          "subDirectory": {},
        },
      },
    },
  },
}
`;

exports[`test > paths to tree 1`] = `
{
  "items": [
    {
      "data": undefined,
      "ext": "",
      "file": ".gitignore",
      "filename": ".gitignore",
      "isEntry": false,
      "parent": [Circular],
      "path": ".gitignore",
    },
  ],
  "name": "root",
  "path": "/",
  "relativePath": "/",
  "relativePathName": "",
  "subDirectory": {
    "a": {
      "items": [
        {
          "data": undefined,
          "ext": "ts",
          "file": "index.ts",
          "filename": "index",
          "isEntry": true,
          "parent": [Circular],
          "path": "a/index.ts",
        },
        {
          "data": undefined,
          "ext": "ts",
          "file": "b.ts",
          "filename": "b",
          "isEntry": false,
          "parent": [Circular],
          "path": "a/b.ts",
        },
      ],
      "name": "a",
      "parent": [Circular],
      "path": "/a/",
      "relativePath": "a/",
      "relativePathName": "a",
      "subDirectory": {
        "b": {
          "items": [
            {
              "data": undefined,
              "ext": "js",
              "file": "index.js",
              "filename": "index",
              "isEntry": true,
              "parent": [Circular],
              "path": "a/b/index.js",
            },
          ],
          "name": "b",
          "parent": [Circular],
          "path": "/a/b/",
          "relativePath": "b/",
          "relativePathName": "b",
          "subDirectory": {},
        },
      },
    },
    "src": {
      "items": [
        {
          "data": undefined,
          "ext": "ts",
          "file": "index.ts",
          "filename": "index",
          "isEntry": true,
          "parent": [Circular],
          "path": "src/index.ts",
        },
        {
          "data": undefined,
          "ext": "ts",
          "file": "foo.ts",
          "filename": "foo",
          "isEntry": false,
          "parent": [Circular],
          "path": "src/foo.ts",
        },
      ],
      "name": "src",
      "parent": [Circular],
      "path": "/src/",
      "relativePath": "src/",
      "relativePathName": "src",
      "subDirectory": {
        "a": {
          "items": [
            {
              "data": undefined,
              "ext": "ts",
              "file": "index.ts",
              "filename": "index",
              "isEntry": true,
              "parent": [Circular],
              "path": "src/a/index.ts",
            },
          ],
          "name": "a",
          "parent": [Circular],
          "path": "/src/a/",
          "relativePath": "a/",
          "relativePathName": "a",
          "subDirectory": {
            "b": {
              "items": [
                {
                  "data": undefined,
                  "ext": "ts",
                  "file": "index.ts",
                  "filename": "index",
                  "isEntry": true,
                  "parent": [Circular],
                  "path": "src/a/b/index.ts",
                },
              ],
              "name": "b",
              "parent": [Circular],
              "path": "/src/a/b/",
              "relativePath": "b/",
              "relativePathName": "b",
              "subDirectory": {
                "d": {
                  "items": [
                    {
                      "data": undefined,
                      "ext": "ts",
                      "file": "index.ts",
                      "filename": "index",
                      "isEntry": true,
                      "parent": [Circular],
                      "path": "src/a/b/d/index.ts",
                    },
                  ],
                  "name": "d",
                  "parent": [Circular],
                  "path": "/src/a/b/d/",
                  "relativePath": "d/",
                  "relativePathName": "d",
                  "subDirectory": {},
                },
              },
            },
          },
        },
        "b": {
          "items": [
            {
              "data": undefined,
              "ext": "js",
              "file": "index.js",
              "filename": "index",
              "isEntry": true,
              "parent": [Circular],
              "path": "src/b/index.js",
            },
          ],
          "name": "b",
          "parent": [Circular],
          "path": "/src/b/",
          "relativePath": "b/",
          "relativePathName": "b",
          "subDirectory": {
            "c": {
              "items": [
                {
                  "data": undefined,
                  "ext": "ts",
                  "file": "index.ts",
                  "filename": "index",
                  "isEntry": true,
                  "parent": [Circular],
                  "path": "src/b/c/index.ts",
                },
              ],
              "name": "c",
              "parent": [Circular],
              "path": "/src/b/c/",
              "relativePath": "c/",
              "relativePathName": "c",
              "subDirectory": {},
            },
          },
        },
      },
    },
  },
}
`;

exports[`test > walker 1`] = `
[
  {
    "data": undefined,
    "ext": "ts",
    "file": "index.ts",
    "filename": "index",
    "isEntry": true,
    "parent": {
      "items": [Circular],
      "name": "b",
      "parent": {
        "items": [
          {
            "data": undefined,
            "ext": "ts",
            "file": "index.ts",
            "filename": "index",
            "isEntry": true,
            "parent": [Circular],
            "path": "src/a/index.ts",
          },
        ],
        "name": "a",
        "parent": {
          "items": [
            {
              "data": undefined,
              "ext": "ts",
              "file": "index.ts",
              "filename": "index",
              "isEntry": true,
              "parent": [Circular],
              "path": "src/index.ts",
            },
            {
              "data": undefined,
              "ext": "ts",
              "file": "foo.ts",
              "filename": "foo",
              "isEntry": false,
              "parent": [Circular],
              "path": "src/foo.ts",
            },
          ],
          "name": "src",
          "parent": {
            "items": [
              {
                "data": undefined,
                "ext": "",
                "file": ".gitignore",
                "filename": ".gitignore",
                "isEntry": false,
                "parent": [Circular],
                "path": ".gitignore",
              },
            ],
            "name": "root",
            "path": "/",
            "relativePath": "/",
            "relativePathName": "",
            "subDirectory": {
              "a": {
                "items": [
                  {
                    "data": undefined,
                    "ext": "ts",
                    "file": "index.ts",
                    "filename": "index",
                    "isEntry": true,
                    "parent": [Circular],
                    "path": "a/index.ts",
                  },
                  {
                    "data": undefined,
                    "ext": "ts",
                    "file": "b.ts",
                    "filename": "b",
                    "isEntry": false,
                    "parent": [Circular],
                    "path": "a/b.ts",
                  },
                ],
                "name": "a",
                "parent": [Circular],
                "path": "/a/",
                "relativePath": "a/",
                "relativePathName": "a",
                "subDirectory": {
                  "b": {
                    "items": [
                      {
                        "data": undefined,
                        "ext": "js",
                        "file": "index.js",
                        "filename": "index",
                        "isEntry": true,
                        "parent": [Circular],
                        "path": "a/b/index.js",
                      },
                    ],
                    "name": "b",
                    "parent": [Circular],
                    "path": "/a/b/",
                    "relativePath": "b/",
                    "relativePathName": "b",
                    "subDirectory": {},
                  },
                },
              },
              "src": [Circular],
            },
          },
          "path": "/src/",
          "relativePath": "src/",
          "relativePathName": "src",
          "subDirectory": {
            "a": [Circular],
            "b": {
              "items": [
                {
                  "data": undefined,
                  "ext": "js",
                  "file": "index.js",
                  "filename": "index",
                  "isEntry": true,
                  "parent": [Circular],
                  "path": "src/b/index.js",
                },
              ],
              "name": "b",
              "parent": [Circular],
              "path": "/src/b/",
              "relativePath": "b/",
              "relativePathName": "b",
              "subDirectory": {
                "c": {
                  "items": [
                    {
                      "data": undefined,
                      "ext": "ts",
                      "file": "index.ts",
                      "filename": "index",
                      "isEntry": true,
                      "parent": [Circular],
                      "path": "src/b/c/index.ts",
                    },
                  ],
                  "name": "c",
                  "parent": [Circular],
                  "path": "/src/b/c/",
                  "relativePath": "c/",
                  "relativePathName": "c",
                  "subDirectory": {},
                },
              },
            },
          },
        },
        "path": "/src/a/",
        "relativePath": "a/",
        "relativePathName": "a",
        "subDirectory": {
          "b": [Circular],
        },
      },
      "path": "/src/a/b/",
      "relativePath": "b/",
      "relativePathName": "b",
      "subDirectory": {
        "d": {
          "items": [
            {
              "data": undefined,
              "ext": "ts",
              "file": "index.ts",
              "filename": "index",
              "isEntry": true,
              "parent": [Circular],
              "path": "src/a/b/d/index.ts",
            },
          ],
          "name": "d",
          "parent": [Circular],
          "path": "/src/a/b/d/",
          "relativePath": "d/",
          "relativePathName": "d",
          "subDirectory": {},
        },
      },
    },
    "path": "src/a/b/index.ts",
  },
]
`;
